### ISSUE TITLE
Cupertino Back swipe gesture causing null-unsafe operations

### ISSUE THREAD

**Issue by mehmetf** [flutter/flutter#63984]
Internal: b/164920328

We have no small repro for this unfortunately.

Two crash reports started coming in from apps that have switched to null-safe code:

```
Thread 0 main thread CRASHED_CastError: Null check operator used on a null value
at AnimationController.stop(animation_controller.dart:747)
at AnimationController._animateToInternal(animation_controller.dart:581)
at AnimationController.animateTo(animation_controller.dart:518)
at _CupertinoBackGestureController.dragEnd(route.dart:773)
at _CupertinoBackGestureDetectorState._handleDragEnd(route.dart:657)
at DragGestureRecognizer._checkEnd.<anonymous closure>(monodrag.dart:440)
at GestureRecognizer.invokeCallback(recognizer.dart:183)
at DragGestureRecognizer._checkEnd(monodrag.dart:440)
at DragGestureRecognizer.didStopTrackingLastPointer(monodrag.dart:349)
at OneSequenceGestureRecognizer.stopTrackingPointer(recognizer.dart:342)
at DragGestureRecognizer._giveUpPointer(monodrag.dart:358)
at DragGestureRecognizer.handleEvent(monodrag.dart:282)
at PointerRouter._dispatch(pointer_router.dart:77)
at PointerRouter._dispatchEventToRoutes.<anonymous closure>(pointer_router.dart:122)
at _LinkedHashMapMixin.forEach(compact_hash.dart:377)
at PointerRouter._dispatchEventToRoutes(pointer_router.dart:120)
at PointerRouter.route(pointer_router.dart:106)
at GestureBinding.handleEvent(binding.dart:219)
at GestureBinding.dispatchEvent(binding.dart:199)
at GestureBinding._handlePointerEvent(binding.dart:157)
at GestureBinding._flushPointerEventQueue(binding.dart:103)
```

and

```
Thread 0 main thread CRASHED_CastError: Null check operator used on a null value
at DragGestureRecognizer.acceptGesture(monodrag.dart:294)
at GestureArenaManager.sweep(arena.dart:157)
at GestureBinding.handleEvent(binding.dart:223)
at GestureBinding.dispatchEvent(binding.dart:199)
at GestureBinding._handlePointerEvent(binding.dart:157)
at GestureBinding._flushPointerEventQueue(binding.dart:103)
at GestureBinding._handlePointerDataPacket(binding.dart:87)
at _rootRunUnary(zone.dart:1206)
at _CustomZone.runUnary(zone.dart:1100)
at _CustomZone.runUnaryGuarded(zone.dart:1005)
at _invoke1(hooks.dart:267)
at _dispatchPointerDataPacket(hooks.dart:176)
```

It seems like gesture handlers are using certain objects beyond their useful lifecycle.

**Comment by HansMuller** [comment]
CC @gspencergoog @goderbauer

**Comment by goderbauer** [comment]
The first crash is happening when a route is removed while the user is in the process of doing a back swipe. It can be reproduced with the following code:

```dart
import 'package:flutter/material.dart';

void main() => runApp(MaterialApp(home: MyApp()));

class MyApp extends StatelessWidget {
  @override
  Widget build(BuildContext context) {
    return Scaffold(
      body: Center(
        child: RaisedButton(
          child: Text('press me'),
          onPressed: () {
            Navigator.of(context).push(r);
            Future.delayed(const Duration(seconds: 5), () {
              Navigator.of(context).removeRoute(r);
            });
          },
        ),
      ),
    );
  }
}

MaterialPageRoute r = MaterialPageRoute(builder: (BuildContext context) {
  return Scaffold(
    body: Center(
      child: Text('gg'),
    ),
  );
});
```

Press the button, start a back-swipe from the left edge and hold it a few pixels from the left edge. Wait until the timer fires ==> crash. This crash doesn't look specific to the NNBD migration. This code would have crashed before as well with a null pointer error.

The problem is that the Route gets disposed in removeRoute, but due to the ongoing back swipe gesture the `_CupertinoBackGestureController` is still holding on to the animation controller of the route and when it tries to interact with it, it crashes because the controller has been disposed with the route.

The second crash looks unrelated to this. Can you file a separate issue for that @mehmetf?

**Comment by goderbauer** [comment]
(we should file a separate bug for the second crash, but) Not sure what code pass can trigger _lastPendingEventTimestamp to be null here:

https://github.com/flutter/flutter/blob/f0dd4ad5302d5424f77094bc50ebb7f2b32c447a/packages/flutter/lib/src/gestures/monodrag.dart#L322

/cc @dkwingsmt Do you have any idea?

**Comment by mehmetf** [comment]
Done: https://github.com/flutter/flutter/issues/64253

**Comment by Hixie** [comment]
@goderbauer can you take this one or assign it appropriately?

**Comment by mehmetf** [comment]
Similar stack trace in a top 10 internal framework crash:

```
Thread 0 main thread CRASHED
NoSuchMethodError
at CupertinoRouteTransitionMixin.isPopGestureInProgress(route.dart:173)
at PageTransitionsTheme.buildTransitions(page_transitions_theme.dart:596)
at MaterialRouteTransitionMixin.buildTransitions(page.dart:124)
at _ModalScopeState.build.<anonymous closure>(routes.dart:794)
at AnimatedBuilder.build(transitions.dart:1252)
at _AnimatedState.build(transitions.dart:181)
at StatefulElement.build(framework.dart:4681)
at ComponentElement.performRebuild(framework.dart:4564)
at StatefulElement.performRebuild(framework.dart:4737)
at Element.rebuild(framework.dart:4280)
at BuildOwner.buildScope(framework.dart:2673)
at WidgetsBinding.drawFrame(binding.dart:913)
at RendererBinding._handlePersistentFrameCallback(binding.dart:298)
at SchedulerBinding._invokeFrameCallback(binding.dart:1117)
at SchedulerBinding.handleDrawFrame(binding.dart:1055)
at SchedulerBinding._handleDrawFrame(binding.dart:971)
at _rootRun(zone.dart:1190)
at _CustomZone.run(zone.dart:1093)
at _CustomZone.runGuarded(zone.dart:997)
at _invoke(hooks.dart:251)
at _drawFrame(hooks.dart:209)
```

**Comment by Hixie** [comment]
This is currently blocked on https://github.com/flutter/flutter/pull/64846.

**Comment by dkwingsmt** [comment]
Update: I've reproduced it in unit test and is working on it right now.

**Comment by dkwingsmt** [comment]
Cause of this issue:
- A route is popped and disposed. During the disposal,
  - Its animation ticker is disposed.
  - The active pointer is canceled by synthesis.
- The disposal of the ticker should trigger the navigator to rebuild, removing all dependents of the ticker, but it does not come until the next frame.
- _Before_ the next frame, the cancel event is dispatched.
  - The drag gesture recognizer (which is not disposed until the next frame) receives the cancel event and starts to dispatch ending callbacks. The callback starts some animation, only to find the ticker disposed, triggering assertion error.

**Comment by mehmetf** [comment]
This is now a P1 for customer:money as it is causing 36% of their crashes during dogfood.

**Comment by mehmetf** [comment]
Since it has been 2 months, please let us know if there's a blocker for this issue.

**Comment by goderbauer** [comment]
@mehmetf There is no blocker, a fix for this issue is waiting for review at https://github.com/flutter/flutter/pull/68913 and should land soon. Sorry for the delay, the issue was a little tricky to debug and fix.
